File: generated_code/GeminiPro/Java/029_SplayTree.java
/**
 * A complete, self-contained implementation of an integer Splay Tree.
 *
 * A splay tree is a self-balancing binary search tree with the additional
 * property that recently accessed elements are quick to access again. It performs
 * basic operations such as insertion, deletion, and search. When an element is
 * accessed (e.g., through search, insertion, or deletion), it is "splayed"
 * to become the new root of the tree.
 *
 * This implementation supports the following public methods for an integer set:
 * - insert(int key): Adds a key to the set. If the key already exists, the
 *   existing node is splayed to the root.
 * - delete(int key): Removes a key from the set.
 * - search(int key): Searches for a key. If found, it returns the key and
 *   splays the found node to the root. If not found, it returns null and splays
 *   the last accessed node to the root.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     * It is an inner class to be used only by the SplayTree class.
     */
    private class Node {
        int key;
        Node parent;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.parent = null;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        root = null;
    }

    /**
     * Performs a left rotation on the given node x.
     *
     *      x
     *     / \
     *    A   y
     *       / \
     *      B   C
     *
     * becomes
     *
     *      y
     *     / \
     *    x   C
     *   / \
     *  A   B
     *
     * @param x The node to rotate.
     */
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y != null) {
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.left = x;
            x.parent = y;
        }
    }

    /**
     * Performs a right rotation on the given node x.
     *
     *        x
     *       / \
     *      y   C
     *     / \
     *    A   B
     *
     * becomes
     *
     *      y
     *     / \
     *    A   x
     *       / \
     *      B   C
     *
     * @param x The node to rotate.
     */
    private void rightRotate(Node x) {
        Node y = x.left;
        if (y != null) {
            x.left = y.right;
            if (y.right != null) {
                y.right.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.right) {
                x.parent.right = y;
            } else {
                x.parent.left = y;
            }
            y.right = x;
            x.parent = y;
        }
    }

    /**
     * Moves the given node to the root of the tree using a series of rotations.
     *
     * @param x The node to splay.
     */
    private void splay(Node x) {
        while (x.parent != null) {
            Node parent = x.parent;
            Node grandParent = parent.parent;
            if (grandParent == null) { // Zig case
                if (x == parent.left) {
                    rightRotate(parent);
                } else {
                    leftRotate(parent);
                }
            } else {
                if (x == parent.left) {
                    if (parent == grandParent.left) { // Zig-Zig case
                        rightRotate(grandParent);
                        rightRotate(parent);
                    } else { // Zig-Zag case
                        rightRotate(parent);
                        leftRotate(grandParent);
                    }
                } else {
                    if (parent == grandParent.left) { // Zig-Zag case
                        leftRotate(parent);
                        rightRotate(grandParent);
                    } else { // Zig-Zig case
                        leftRotate(grandParent);
                        leftRotate(parent);
                    }
                }
            }
        }
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the node with that key is splayed to the root.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        Node node = new Node(key);
        Node parent = null;
        Node current = this.root;

        while (current != null) {
            parent = current;
            if (node.key < current.key) {
                current = current.left;
            } else if (node.key > current.key) {
                current = current.right;
            } else {
                // Key already exists, splay the found node and return
                splay(current);
                return;
            }
        }

        node.parent = parent;
        if (parent == null) {
            root = node; // Tree was empty
        } else if (node.key < parent.key) {
            parent.left = node;
        } else {
            parent.right = node;
        }

        // Splay the newly inserted node
        splay(node);
    }

    /**
     * Searches for a key in the splay tree.
     * If the key is found, the corresponding node is splayed to the root and the
     * key is returned.
     * If the key is not found, the last accessed node on the search path is
     * splayed to the root and null is returned.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }

        Node current = root;
        Node lastNode = root;

        while (current != null) {
            lastNode = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key found
                splay(current);
                return current.key;
            }
        }

        // Key not found, splay the last visited node
        splay(lastNode);

        // After splaying, the new root might be the key if it was found,
        // but our loop logic already handled that. If we are here, key was not found.
        return null;
    }

    /**
     * Deletes a key from the splay tree.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        // First, search for the key. This splays the node (or its neighbor) to the root.
        search(key);

        // If root is null or key is not at the root, the key wasn't in the tree.
        if (root == null || root.key != key) {
            return;
        }

        Node nodeToDelete = root;
        Node leftSubtree = nodeToDelete.left;
        Node rightSubtree = nodeToDelete.right;

        // Case 1: No left child. The right child becomes the new root.
        if (leftSubtree == null) {
            root = rightSubtree;
            if (root != null) {
                root.parent = null;
            }
        }
        // Case 2: Has a left child.
        else {
            // Detach the left subtree to work on it independently
            leftSubtree.parent = null;

            // Find the maximum node in the left subtree
            Node maxNode = leftSubtree;
            while (maxNode.right != null) {
                maxNode = maxNode.right;
            }

            // Temporarily make the left subtree the main tree to splay within it
            this.root = leftSubtree;
            splay(maxNode); // maxNode is now the root of the (modified) left subtree

            // The new root of the entire tree is this splayed maxNode
            // Attach the original right subtree
            this.root.right = rightSubtree;
            if (rightSubtree != null) {
                rightSubtree.parent = this.root;
            }
        }
    }
}

File: generated_code/GeminiPro/Python/029_SplayTree.py
import sys

class SplayTree:
    """
    A self-contained Python class implementing a Splay Tree.
    This tree acts as a dictionary-like set for storing unique integer keys.
    """

    class _Node:
        """A node in the Splay Tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x
        
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x

        y.parent = x.parent
        if x.parent is None:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
            
        y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Performs the splaying operation on a node, moving it to the root.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:
                # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:
                # Zig-Zig case (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:
                # Zig-Zig case (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:
                # Zig-Zag case (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:
                # Zig-Zag case (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.
        
        If the key is found, the corresponding node is splayed to the root.
        If the key is not found, the last accessed node (the would-be parent)
        is splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        last_visited = None
        node = self.root
        while node:
            last_visited = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        
        # If key not found, splay the last visited node
        if last_visited:
            self._splay(last_visited)
        
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.
        
        If the key is not present, it is added and splayed to the root.
        If the key already exists, the existing node is splayed to the root.

        Args:
            key: The integer key to insert.
        """
        parent = None
        current = self.root
        
        # Find position for new node or existing node
        while current:
            parent = current
            if key == current.key:
                # Key already exists, splay it and return
                self._splay(current)
                return
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # Insert new node
        new_node = self._Node(key)
        new_node.parent = parent
        
        if not parent:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
            
        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.
        
        The tree is first splayed on the key. If the key exists, it is
        removed, and its two subtrees are joined.

        Args:
            key: The integer key to delete.
        """
        # Splay the node to the root (or its parent if not found)
        if not self.search(key):
            # Key was not in the tree, search already splayed the parent
            return

        # At this point, the node to delete is the root, because search(key)
        # returned True, splaying the node with the given key to the root.
        z = self.root
        left_subtree = z.left
        right_subtree = z.right

        if not left_subtree:
            # No left child, promote the right subtree
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Disconnect the left subtree to operate on it
            left_subtree.parent = None
            
            # Find the maximum node in the left subtree
            max_in_left = left_subtree
            while max_in_left.right:
                max_in_left = max_in_left.right
            
            # Splay this maximum node to the root of the left subtree
            # We can do this by temporarily setting self.root and using _splay
            temp_root_holder = self.root
            self.root = left_subtree
            self._splay(max_in_left)
            # After splaying, self.root is now max_in_left
            
            # The new root is the splayed max node from the left subtree
            new_root = self.root
            self.root = temp_root_holder # Restore original root context (not strictly needed)

            # Join the right subtree to the new root
            new_root.right = right_subtree
            if right_subtree:
                right_subtree.parent = new_root
            
            self.root = new_root
            self.root.parent = None
    
    # Helper for visualization
    def _print_helper(self, node, prefix="", is_left=True):
        if not node:
            return
        if node.right:
            self._print_helper(node.right, prefix + ("│   " if is_left else "    "), False)
        print(prefix + ("└── " if is_left else "┌── ") + str(node.key))
        if node.left:
            self._print_helper(node.left, prefix + ("    " if is_left else "│   "), True)

    def print_tree(self):
        """Prints a visual representation of the tree."""
        if self.root:
            self._print_helper(self.root)
        else:
            print("Tree is empty.")


if __name__ == '__main__':
    # Example Usage
    splay_tree = SplayTree()

    print("Inserting 10, 20, 30, 40, 50, 5, 15...")
    keys_to_insert = [10, 20, 30, 40, 50, 5, 15]
    for key in keys_to_insert:
        splay_tree.insert(key)
        print(f"\nAfter inserting {key}, root is {splay_tree.root.key}:")
        splay_tree.print_tree()

    print("\n" + "="*40)
    print("Searching for key 15...")
    found = splay_tree.search(15)
    print(f"Key 15 found: {found}")
    print(f"Tree after searching for 15 (root is {splay_tree.root.key}):")
    splay_tree.print_tree()

    print("\n" + "="*40)
    print("Searching for key 99 (not in tree)...")
    found = splay_tree.search(99)
    print(f"Key 99 found: {found}")
    print(f"Tree after searching for 99 (root is {splay_tree.root.key}, the last visited node):")
    splay_tree.print_tree()

    print("\n" + "="*40)
    print("Deleting key 30...")
    splay_tree.delete(30)
    print(f"Tree after deleting 30 (root is {splay_tree.root.key}):")
    splay_tree.print_tree()
    
    print("\n" + "="*40)
    print("Deleting key 50...")
    splay_tree.delete(50)
    print(f"Tree after deleting 50 (root is {splay_tree.root.key}):")
    splay_tree.print_tree()

    print("\n" + "="*40)
    print("Deleting key 99 (not in tree)...")
    splay_tree.delete(99)
    print(f"Tree after attempting to delete 99 (root is {splay_tree.root.key}):")
    splay_tree.print_tree()